Added support for named params with regex path

vigenere23 opened the pull request:
fixes #154 (partially)

This PR is a proposal for adding support for name params (aka `:id` and etc.) while using `RegExp` paths. 

> Please note that this is a proposal for code example and is clearly not complete

### How it works

`regexparam` handles `RegExp` paths quite weirdly and returns `false` as `keys` instead of the actual wanted keys (which works well when `path` is a `string`). To make it work, we have to:

1. manually iterate over the string value of the `RegExp` path
2. find all occurences of named groups (starting with `?<` and ending with `>`)
3. adding the found named group in the path, minus the delimiters
4. replace that part of the path by adding an almost-wildcard pattern `[^\/]+?` just like it does for string path params (`:param`)
5. since this string comes from `RegExp.prototype.toString()`, it includes the `/` delimiters + the flag, which **should not** be present when constructing a new `RegExp`, so they are removed

### Not done in PR

- Tests
- Edge cases
- Optional params

ItalyPaleAle commented:
Thanks for the idea and for submitting a PR! I like the idea.

I'm on vacation without my laptop for the next few days so I'll need some time for reviewing this better. At a glance, however, it seems ok.

The only thing that makes me a little bit uncomfortable is that there are 3 regular expressions for parsing the regexp. Implementing a Parser with a regexp is generally not the best thing, so I wonder if that could be rewritten to be an actual parser? Like: removing all 3 regular expressions chained and using a for loop instead to do the parsing.

vigenere23 commented:
> I'm on vacation without my laptop for the next few days so I'll need some time for reviewing this better. At a glance, however, it seems ok.

Have fun! :beach_umbrella: 

> The only thing that makes me a little bit uncomfortable is that there are 3 regular expressions for parsing the regexp. Implementing a Parser with a regexp is generally not the best thing, so I wonder if that could be rewritten to be an actual parser? Like: removing all 3 regular expressions chained and using a for loop instead to do the parsing.

True, some of those where clearly not needed. I've replaced the last 2 ones with simple `.slice()`s. Should I try to replace the one that checks for path params (`?<param>`) too? This one may be quite harder/cumbersome to do :sweat_smile: 

I also extracted the whole thing to a function and added some documentation and explanations :slightly_smiling_face:
